src/machine/asm/mod.rs

```diff
@@ -101,6 +102,97 @@ fn check_memory(machine: &mut AsmCoreMachine, page_indices: &(u64, u64)) {
     }
 }
 
+// check whether a memory address is writable or not and mark it as dirty, `size` should be 1, 2, 4 or 8
+fn check_memory_writable(
+    machine: &mut Box<AsmCoreMachine>,
+    addr: u64,
+    size: usize,
+) -> Result<(), Error> {
+    debug_assert!(size == 1 || size == 2 || size == 4 || size == 8);
+    let page = addr >> RISCV_PAGE_SHIFTS;
+    if page as usize >= RISCV_PAGESIZE {
+        return Err(Error::OutOfBound);
+    }
+
+    let page_offset = addr as usize % RISCV_PAGESIZE;
```

Comment:
May be faster if the compiler does not do this optimization : `let page_offset = addr as usize >> RISCV_PAGE_SHIFTS`

Reply:
I can confirm that the compiler will optimize the code, will keep it unchanged